### ISSUE TITLE
Symmetry breaks quickly in simple flow

### ISSUE THREAD

**Issue by mcgratta** [firemodels/fds#13764]
Run this case below. It only takes a few seconds. The resulting flow should be symmetric and isothermal, but it is neither. It is worst when `CONSTANT_SPECIFIC_HEAT_RATIO=T`, but it still fails if is `F`. If I remove one of the `VENTs` injecting the gas, the case remains iso-thermal, but obviously not symmetric.
```
&HEAD CHID='sym_test' /

&MESH IJK=64,1,16, XB=0.0,3.2,-0.001,0.001,0.0,0.8 /

&TIME T_END=8.0 /

&PRES CHECK_POISSON=T /

&MISC STRATIFICATION=F, NOISE=F, CONSTANT_SPECIFIC_HEAT_RATIO=T /
&RADI RADIATION=F /

&SPEC ID='GAS BG', MW=29, BACKGROUND=T /
&SPEC ID='GAS A', MW=29 /
&SPEC ID='GAS B', MW=29 /

&SURF ID='WALL', FREE_SLIP=T, DEFAULT=.TRUE. /

&SURF ID='BLOW A', VEL=-0.5, MASS_FRACTION(1)=1., SPEC_ID(1)='GAS A' /
&SURF ID='BLOW B', VEL=-0.5, MASS_FRACTION(1)=1., SPEC_ID(1)='GAS B' /

&VENT XB=0.0,0.0,-0.001,0.001,0.0,0.4, SURF_ID='BLOW A' /
&VENT XB=0.0,0.0,-0.001,0.001,0.4,0.8, SURF_ID='BLOW B' /
&VENT PBX=3.2, SURF_ID='OPEN' /

&SLCF PBY=0.,QUANTITY='MASS FRACTION', SPEC_ID='GAS A',  CELL_CENTERED=T, VECTOR=T /
&SLCF PBY=0.,QUANTITY='MASS FRACTION', SPEC_ID='GAS B',  CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='MASS FRACTION', SPEC_ID='GAS BG', CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='TEMPERATURE', CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='DIVERGENCE',  CELL_CENTERED=T /

&DEVC ID='T_max', XB=0.0,3.2,-0.001,0.001,0.0,0.8, QUANTITY='TEMPERATURE', SPATIAL_STATISTIC='MAX', TEMPORAL_STATISTIC='INSTANT VALUE' /
&DEVC ID='T_min', XB=0.0,3.2,-0.001,0.001,0.0,0.8, QUANTITY='TEMPERATURE', SPATIAL_STATISTIC='MIN', TEMPORAL_STATISTIC='INSTANT VALUE' /

&DUMP DT_DEVC=0.00001 /

&TAIL /
```

**Comment by drjfloyd** [comment]
With a quick look. The .out file show these all have the same properties and if I add slices for specific heat, conductivity, etc. Those show identical values until temperature starts to change in which case conductivity increases with temperature. This doesn't seem to be something wrong with the functions for getting properties.

**Comment by mcgratta** [comment]
I don't think so either. I'm just surprised at how quickly the symmetry and iso-thermal break down. Like in 3 time steps. No time step changes, no changes in geom, no obsts, no mesh boundaries.

**Comment by drjfloyd** [comment]
It also doesn't happen if one of the VENTs is the background species or both vents are the same species. Only when the two vents are A and B

**Comment by mcgratta** [comment]
Nice clue.

**Comment by mcgratta** [comment]
How did you apply the BG species at the VENT. I get
```
ERROR(330): SURF BLOW B cannot use background species for MASS_FRACTION. (CHID: sym_test)
```

**Comment by mcgratta** [comment]
Nevermind. I know how.

**Comment by mcgratta** [comment]
Setting `GVEC=0,0,0` maintains symmetry even though I would think we would not have to set that to do so.

**Comment by mcgratta** [comment]
When I run the case longer, the initial slop goes away. There seems to be something about the ramp up or initial/boundary conditions.
![image](https://github.com/user-attachments/assets/c019ce46-8846-4a06-bafc-dce300916738)

**Comment by drjfloyd** [comment]
It is still seen with just a 2 cell wide domain but not as strong

&HEAD CHID='test' /

&MESH IJK=8,1,4, XB=0.0,0.8,-0.001,0.001,0.2,0.6 /

&TIME T_END=2,DT=0.01 /

&PRES CHECK_POISSON=T /

&MISC STRATIFICATION=F, NOISE=F, CONSTANT_SPECIFIC_HEAT_RATIO=T /
&RADI RADIATION=F /

&SPEC ID='GAS BG', MW=29, BACKGROUND=T /
&SPEC ID='GAS A', MW=29 /
&SPEC ID='GAS B', MW=29 /

&SURF ID='WALL', FREE_SLIP=T, DEFAULT=.TRUE. /

&SURF ID='BLOW A', VEL=-0.5, MASS_FRACTION(1)=1., SPEC_ID(1)='GAS A' /
&SURF ID='BLOW B', VEL=-0.5, MASS_FRACTION(1)=1., SPEC_ID(1)='GAS B' /

&VENT XB=0.0,0.0,-0.001,0.001,0.3,0.4, SURF_ID='BLOW A' /
&VENT XB=0.0,0.0,-0.001,0.001,0.4,0.5, SURF_ID='BLOW B' /

&OBST XB=0,0.8,-0.001,0.001,0.2,0.3/
&OBST XB=0,0.8,-0.001,0.001,0.5,0.6/

&VENT PBX=0.8, SURF_ID='OPEN' /

&SLCF PBY=0.,QUANTITY='MASS FRACTION', SPEC_ID='GAS A',  CELL_CENTERED=T, VECTOR=T /
&SLCF PBY=0.,QUANTITY='MASS FRACTION', SPEC_ID='GAS B',  CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='MASS FRACTION', SPEC_ID='GAS BG', CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='TEMPERATURE', CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='DIVERGENCE',  CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='CONDUCTIVITY',  CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='SPECIFIC HEAT',  CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='DENSITY',  CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='VISCOSITY',  CELL_CENTERED=T /
&SLCF PBY=0.,QUANTITY='SPECIFIC ENTHALPY',  CELL_CENTERED=T /

&TAIL/

![test_0175](https://github.com/user-attachments/assets/1d621da8-e774-485a-9789-45c63529ac8e)

**Comment by mcgratta** [comment]
On a related note, I just added a case called `Pressure_Solver/obst_activation_default_gases.fds` which is what this issue is based on. In this case, there are multiple meshes and obstructions that are crated and removed at mesh boundaries. I set the flux limiter to `GODUNOV` because, for the moment, I am trying to fix a small discrepancy when obstructions are created or removed at mesh boundaries in the presence of mixing gases. I think that this case be dealt with separately form the flux limiter issue.

**Comment by rmcdermo** [comment]
OK, I should be able to track this down.  The other limiters should work too, so something is amiss.

**Comment by mcgratta** [comment]
In the case where we have three gases, two on the inlet and one background, and all have the same MW, when I sum the limited fluxes, I do not get what I expect with `SUPERBEE`, i.e. $\sum_\alpha \overline{\rho Z}_\alpha^{\rm FL} = \rho$, but with `GODUNOV` I do. I have not worked out the algebra to demonstrate that I should or should not expect this.

**Comment by rmcdermo** [comment]
I have a guess at what is going on.  I sent some notes on Overleaf (they were not rendering correctly here on GitHub).

**Comment by rmcdermo** [comment]
I added a correction in this PR #13789 .  For now, it is hidden behind the MISC flag `TEST_FLUX_LIMITER_FACE_CORRECTION`.  The following input file now gives perfectly isothermal flow.

[sym_test.fds.txt](https://github.com/user-attachments/files/17852370/sym_test.fds.txt)

<img width="752" alt="Screenshot 2024-11-21 at 4 40 01 PM" src="https://github.com/user-attachments/assets/f16ee3a2-fa1b-4bb0-8b1b-c3137398b83b">

Once we are happy with the coding, we can remove the flag.  I think we just need to invoke this if N_TRACKED_SPECIES>2.

**Comment by drjfloyd** [comment]
Should we set the flag to TRUE for tonight's firebot?

**Comment by rmcdermo** [comment]
Sure.  I would be ok with that.  I did not want to throw another grenade in with all the other failures.  But if you are reasonably happy with the routine then we can give it a try.  Kevin, please weigh in.

**Comment by mcgratta** [comment]
Too many communication channels. Let me take a look.

**Comment by mcgratta** [comment]
I'll run this locally. I'd rather not commit a potential grenade because it causes problems when doing a git bisect.

**Comment by mcgratta** [comment]
I am running the verification cases to see if something will blow up. I set the new parameter to T in the test. But I just ran the `sym_test` case that is posted above, and I gave Gas A a MW of 25 and Gas B a MW of 35. The temperature does not remain isothermal.

![image](https://github.com/user-attachments/assets/b393bce0-457d-436e-b8e7-195aff4ef496)

We can work on this more tomorrow. I think we need to run the idea through a series of increasingly complex cases -- multi-mesh, different gases, flux limiters, obstructions, etc.

**Comment by rmcdermo** [comment]
I think I know what is going on.  I updated the notes.  To maintain isothermal flow, we need to keep $T = \frac{\overline{W}p}{\rho R}$ constant at all cells and cell faces.  The current correction does not consider this.  So, I think I need to add a molecular weight factor such that $\sum_\alpha \frac{(\rho Z_\alpha)}{W_\alpha} = \frac{\rho}{\overline{W}}$ is maintained at the face.

**Comment by mcgratta** [comment]
PR #13789 with the test flag set to T knocked about a dozen cases out of tolerance. Since there is no immediate crisis, I suggest we revert the change and work it over the next few weeks. That will give us time to progress from the simple cases to the complex, like `couch.fds` which was one of the cases that failed.  In the meantime, I want to figure out what is happening at mesh boundaries when obstructions are created or removed  This can be done using the simple `GODUNOV` limiter because I think the issue there is separate. It would be easier not to have the test code complicating things until it's ready.

**Comment by rmcdermo** [comment]
Please do not revert.  I will deal with this.  I am working on the
correction.  Please just leave the flag to false.

**Comment by mcgratta** [comment]
I transferred this issue to Randy. The best test case for this is in `/Verification/Pressure_Solver/obst_activation_default_gases.fds`. You would just need to remove the flag for `GODUNOV`.

**Comment by rmcdermo** [comment]
I made a bit of progress on this issue today.  This PR #13854 is working for this test case:

[isothermal_3spec_const_gamma.fds.txt](https://github.com/user-attachments/files/18044164/isothermal_3spec_const_gamma.fds.txt)

However, this case is single mesh.  When I go to multi-mesh I end up with errors at the INTERPOLATED_BC which is what I'm working on now.

**Comment by drjfloyd** [comment]
Here is a test case. The attached case defines five species that allow for testing with the same MW and CP or varying MW and CP.  Eight meshes have a two cell high gas phase domain contracting down to one cell high. Pairs of species are injected on the left at the same or varied temperature.  For varying temperature comments in the input file note what the solution should be.  

[test.txt](https://github.com/user-attachments/files/18064070/test.txt)

**Comment by rmcdermo** [comment]
This PR #13863 fixes the temperature issue at interpolated boundaries for CONSTANT_SPECIFIC_HEAT_RATIO=T (still).  I will now start working the variable CP aspects using Jason's case.

Another note is that these cases still require CFL_VELOCITY_NORM=1 and CFL_MAX=0.8 to stay perfectly isothermal.  Here is the working input file for `obst_activation_default_gases`.

[test.fds.txt](https://github.com/user-attachments/files/18064175/test.fds.txt)

<img width="400" alt="Screenshot 2024-12-09 at 10 46 52 AM" src="https://github.com/user-attachments/assets/020d8d06-f838-4ae3-8ae6-5d2d1c14482c">

**Comment by rmcdermo** [comment]
@drjfloyd Have your case working in a branch.  Need to run it through firebot then I'll push it up.

<img width="400" alt="Screenshot 2024-12-10 at 11 28 22 AM" src="https://github.com/user-attachments/assets/471067de-f9a7-42f1-9447-ea11cccddd6d">

<img width="400" alt="Screenshot 2024-12-10 at 11 28 28 AM" src="https://github.com/user-attachments/assets/c779e3f0-5046-4eec-a795-086727547809">

**Comment by rmcdermo** [comment]
This should be solved now with `FLUX_LIMITER_MW_CORRECTION=T`, which is now default for LES, PR #13891 

I have pushed this also the verification case here PR #13893 .

I'll close this, but we can reopen if we see issues.  I think that improving efficiency is something that needs its own thread.

**Comment by drjfloyd** [comment]
I just ran firebot with the current source with and without the flux correction set to true.  The symmetry_test_2 case in flowfields failed with it true.  The plot shows the difference in left and right temperature. With it true the difference is almost a degree vs. .03 degree with it false.

![image](https://github.com/user-attachments/assets/f86752e6-ae7d-41b9-a267-c574b64499d8)

**Comment by rmcdermo** [comment]
OK, thanks. I'll take a look.

**Comment by rmcdermo** [comment]
@drjfloyd I'm not sure this is something we can avoid.  If I set `FLUX_LIMITER_MW_CORRECTION=F` and run the case to 8 seconds, then I get this,

<img width="912" alt="Screenshot 2024-12-20 at 12 04 05 PM" src="https://github.com/user-attachments/assets/0a7709f1-a073-4e57-8f4b-09e8eb8dd233" />

So, this is what we get by default.

If I set,
```
&MISC NOISE=.FALSE., FLUX_LIMITER_MW_CORRECTION=T, CFL_VELOCITY_NORM=1 /
```
then I get this,

<img width="912" alt="Screenshot 2024-12-20 at 12 55 06 PM" src="https://github.com/user-attachments/assets/e64f1bb3-bdbd-47c6-b642-8ecbb5abb5c8" />

I attribute the earlier onset of the randomness simply to the fact that the correction requires more flops and so the roundoff error accumulates a bit faster.

What I can do is to automatically set `CFL_VELOCITY_NORM=1` if `FLUX_LIMITER_MW_CORRECTION=T`.

I don't think it makes sense to try to improve this at the moment, as we have another idea we want to pursue that would both eliminate the need for a special correction and hopefully also speed up the whole transport algorithm.  But let me know if you disagree.
